### PR TITLE
red light violation

### PR DESCRIPTION
**Pull request by emailweixu**

### CODE REVIEW

File: alf/environments/carla_sensors.py [hunk 1]
```diff
@@ -617,9 +618,176 @@ def get_current_observation(self, current_frame):
...
+    c = np.cos(rotation)
+    s = np.sin(rotation)
+    return np.stack([c[:, 1] * c[:, 0], s[:, 1] * c[:, 0], s[:, 0]], axis=-1)
+
+
+def _rotate_point(point: carla.Vector3D, angle):
+    """Rotate a given point by a given angle
+
+    Args:
+        point (carla.Location):
```
Comment (Haichao-Zhang) [inline alf/environments/carla_sensors.py]: Is ```carla.Location``` the same as ```carla.Vector3D```?

File: alf/environments/suite_carla.py [hunk 2]
```diff
@@ -196,14 +204,21 @@ def __init__(self,
             max_collision_penalty (float): the maximum penalty (i.e. negative reward)
                 for collision. We don't want the collision penalty to be too large
                 if the player cannot even get enough positive moving reward. So the
-                panalty is capped at ``max(0., episode_reward))``. Note that this
-                reward is only given once at the first step of contiguous collisions.
+                panalty is capped at ``Player.PENALTY_RATE_COLLISION * max(0., episode_reward))``.
```
Comment (Haichao-Zhang) [inline alf/environments/suite_carla.py]: panalty -> penalty

File: alf/environments/suite_carla.py [hunk 3]
```diff
@@ -196,14 +204,21 @@ def __init__(self,
...
             max_stuck_at_collision_seconds (float): the episode will end and is
                 considerred as failure if the car is stuck at the collision for
                 so many seconds,
             stuck_at_collision_distance (float): the car is considerred as being
                 stuck at the collision if it is within such distance of the first
                 collision location.
+            max_red_light_penalty (float): the maximum penalty (i.e. negative reward)
+                for red light violation. We don't want the collision penalty to
+                be too large if the player cannot even get enough positive moving
+                reward. So the panalty is capped at ``Player.PENALTY_RATE_RED_LIGHT * max(0., episode_reward))``.
```
Comment (Haichao-Zhang) [inline alf/environments/suite_carla.py]: panalty -> penalty

File: alf/environments/carla_sensors.py [hunk 4]
```diff
@@ -617,9 +618,176 @@ def get_current_observation(self, current_frame):
...
+    """
+    x_ = math.cos(math.radians(angle)) * point.x - math.sin(
+        math.radians(angle)) * point.y
+    y_ = math.sin(math.radians(angle)) * point.x + math.cos(
+        math.radians(angle)) * point.y
+    return carla.Vector3D(x_, y_, point.z)
+
+
+def _rotate_np_point(point, angle):
+    """Rotate a given point by a given angle
```
Comment (hnyu) [inline alf/environments/carla_sensors.py:719]: It seems that this function can be reused by _rotate_point above

File: alf/environments/carla_sensors.py [hunk 5]
```diff
@@ -617,9 +618,176 @@ def get_current_observation(self, current_frame):
...
+
+
+def _is_segments_intersecting(seg1, seg2):
+    """Check whether two batched segments intersect.
+
+    Based on the answer by Norbu Tsering at
+    https://stackoverflow.com/questions/3252194/numpy-and-line-intersections
+
+    Args:
+        seg1 (tuple[ndarray]): ``seg1[0]`` and seg1[1] are [1, 3] or [n,3] arrays.
```
Comment (hnyu) [inline alf/environments/carla_sensors.py]: double quotes

File: alf/environments/carla_sensors.py [hunk 6]
```diff
@@ -617,9 +618,176 @@ def get_current_observation(self, current_frame):
...
+    right_lane_marking=dummy_lane_marking,
+    left_lane_marking=dummy_lane_marking,
+)
+
+
 class World(object):
     """Keeping data for the world."""
 
+    # only consider a car for running red light if it is within such distance
+    DISTANCE_LIGHT = 10  # m
```
Comment (hnyu) [inline alf/environments/carla_sensors.py]: Could be a more informative name such as RED_LIGHT_ENFORCE_DISTANCE

File: alf/environments/carla_sensors.py [hunk 7]
```diff
@@ -705,6 +874,164 @@ def transform_to_geolocation(self, location: carla.Location):
...
 
+    def _prepare_traffic_light_data(self):
+        # Adapted from RunningRedLightTest.__init__() in
+        # https://github.com/carla-simulator/scenario_runner/blob/master/srunner/scenariomanager/scenarioatomics/atomic_criteria.py
+
+        actors = self._world.get_actors()
+        self._traffic_light_actors = []
+        traffic_light_centers = []
+        traffic_light_waypoints = []
+        max_num_traffic_light_waypoints = 0
```
Comment (hnyu) [inline alf/environments/carla_sensors.py:887]: Comment on what traffic light waypoints mean

File: alf/environments/carla_sensors.py [hunk 8]
```diff
@@ -705,6 +874,164 @@ def transform_to_geolocation(self, location: carla.Location):
...
+                else:
+                    break
+            wps.append(wpx)
+
+        # self._draw_waypoints(wps, vertical_shift=1.0, persistency=50000.0)
+
+        return area_loc, wps
+
+    def is_running_red_light(self, actor):
+        """Whether actor is running red light.
```
Comment (hnyu) [inline alf/environments/carla_sensors.py:965]: docstring is broken

File: alf/environments/carla_sensors.py [hunk 9]
```diff
@@ -705,6 +874,164 @@ def transform_to_geolocation(self, location: carla.Location):
...
+            d = np.stack([
+                0.4 * lane_width,
+                np.zeros_like(lane_width), location_wp[:, 2]
+            ],
+                         axis=-1)
+            left_lane_wp = _rotate_np_point(d, yaw_wp + 0.5 * math.pi)
+            left_lane_wp = location_wp + left_lane_wp
+            right_lane_wp = _rotate_np_point(d, yaw_wp - 0.5 * math.pi)
+            right_lane_wp = location_wp + right_lane_wp
+            if np.any(same_lane & _is_segments_intersecting(
```
Comment (hnyu) [inline alf/environments/carla_sensors.py:1022]: Comment on what _is_segments_intersecting() checks here?

File: alf/environments/carla_sensors.py [hunk 10]
```diff
@@ -705,6 +874,164 @@ def transform_to_geolocation(self, location: carla.Location):
...
+                    veh_seg, (left_lane_wp, right_lane_wp))):
+                return self._traffic_light_actors[index].id
+        return None
+
+    def _draw_waypoints(self, waypoints, vertical_shift, persistency=-1):
```
Comment (hnyu) [inline alf/environments/carla_sensors.py:1030]: This function seems never used

File: alf/environments/suite_carla.py [hunk 11]
```diff
@@ -196,14 +204,21 @@ def __init__(self,
...
+                Note that this reward is only given once at the first step of
+                contiguous collisions.
             max_stuck_at_collision_seconds (float): the episode will end and is
                 considerred as failure if the car is stuck at the collision for
                 so many seconds,
             stuck_at_collision_distance (float): the car is considerred as being
                 stuck at the collision if it is within such distance of the first
                 collision location.
+            max_red_light_penalty (float): the maximum penalty (i.e. negative reward)
+                for red light violation. We don't want the collision penalty to
```
Comment (hnyu) [inline alf/environments/suite_carla.py]: collision penalty -> red light penalty

File: alf/environments/suite_carla.py (hunk 3 above)
Comment (hnyu) [inline alf/environments/suite_carla.py]: panalty -> penalty

File: alf/environments/suite_carla.py (hunk 2 above)
Comment (hnyu) [inline alf/environments/suite_carla.py]: panalty

File: alf/environments/suite_carla.py [hunk 12]
```diff
@@ -569,6 +590,16 @@ def get_current_time_step(self, current_frame):
                 self._unrecorded_distance_reward = 0
         reward += distance_reward
 
+        red_light_id = self._alf_world.is_running_red_light(self._actor)
+        if red_light_id is not None and red_light_id != self._prev_violated_red_light_id:
+            logging.info("actor=%d frame=%d RED_LIGHT" % (self._actor.id,
+                                                          current_frame))
+            reward_vector[Player.REWARD_RED_LIGHT] = 1.
```
Comment (hnyu) [inline alf/environments/suite_carla.py:597]: Why assign 1 here? Should it be a negative value?

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/environments/carla_sensors.py (hunk 1 above)
Comment (emailweixu) [reply alf/environments/carla_sensors.py]: changed to carla.Vector3D

File: alf/environments/carla_sensors.py (hunk 4 above)
Comment (emailweixu) [reply alf/environments/carla_sensors.py:719]: It's quite a hassle. Need to first convert `point` and `angle` to carla object and convert the result to np.ndarray. In addition to that, _rotate_point cannot handle batched point and angle.

File: alf/environments/carla_sensors.py (hunk 5 above)
Comment (emailweixu) [reply alf/environments/carla_sensors.py]: updated

File: alf/environments/carla_sensors.py (hunk 6 above)
Comment (emailweixu) [reply alf/environments/carla_sensors.py]: changed. It's from the carla code.

File: alf/environments/carla_sensors.py (hunk 7 above)
Comment (emailweixu) [reply alf/environments/carla_sensors.py:887]: commented.

File: alf/environments/carla_sensors.py (hunk 8 above)
Comment (emailweixu) [reply alf/environments/carla_sensors.py:965]: Why did you say it's broken?

File: alf/environments/carla_sensors.py (hunk 9 above)
Comment (emailweixu) [reply alf/environments/carla_sensors.py:1022]: commented

File: alf/environments/carla_sensors.py (hunk 10 above)
Comment (emailweixu) [reply alf/environments/carla_sensors.py:1030]: It's for debugging purpose to visualize.

File: alf/environments/suite_carla.py (hunk 3 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py]: fixed

File: alf/environments/suite_carla.py (hunk 11 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py]: fixed

File: alf/environments/suite_carla.py (hunk 3 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py]: fixed

File: alf/environments/suite_carla.py (hunk 12 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py:597]: Currently, this is just for indicating the event. So it does not matter. And it is different from the actual reward for running red light

Comment (hnyu) [reply alf/environments/suite_carla.py:597]: From my understanding, if we learn with multiple rewards, then there are multiple values learned. So basically the learned value of collision or red light is never used (except display)?

File: alf/environments/carla_sensors.py [hunk 13]
```diff
@@ -1019,6 +1021,9 @@ def is_running_red_light(self, actor):
             right_lane_wp = location_wp + right_lane_wp
             if np.any(same_lane & _is_segments_intersecting(
                     veh_seg, (left_lane_wp, right_lane_wp))):
+                # If veh_seg intersets with (left_lane_wp, right_lane_wp), that
```
Comment (Haichao-Zhang) [inline alf/environments/carla_sensors.py]: intersets  -> intersects

File: alf/environments/suite_carla.py (hunk 12 above)
Comment (emailweixu) [reply alf/environments/suite_carla.py:597]: Right, it's not used. Though it can help to shape the representation.

**Review by Haichao-Zhang — APPROVED** [review]
LGTM

File: alf/environments/carla_sensors.py (hunk 8 above)
Comment (hnyu) [reply alf/environments/carla_sensors.py:965]: Oh I mean the line below has incorrect format.

**Review by hnyu — APPROVED** [review]

File: alf/environments/carla_sensors.py (hunk 8 above)
Comment (emailweixu) [reply alf/environments/carla_sensors.py:965]: Did you mean that the URL is broken into two lines? That's just because github has to wrap the long line.

Comment (hnyu) [reply alf/environments/carla_sensors.py:965]: never mind, I was talking about the code and url format.

**Pull request merged**